### PR TITLE
feat: build download filenames from base filename and version (#1115)

### LINKED ISSUE #1115: Slice 8: Versioned Download Filenames
## Goal

When users download source dataset or integrated object files, the downloaded filename should include version information rather than using the raw S3 key filename. This gives users meaningful, identifiable filenames that reflect the atlas revision and draft state.

## Background

Currently, `getDownloadUrl` extracts the filename from the S3 key's last path segment (e.g. a UUID like `a3f2b1c4-...-.h5ad`) and sets it as the `Content-Disposition` filename on the presigned URL. The concept table already stores a human-readable `base_filename` (e.g. `cells.h5ad`), and the metadata entities (source_datasets, component_atlases) store `wip_number`. The atlas stores `revision`.

## Naming Convention

- **Published** (`published_at IS NOT NULL`): `{name}-r{revision}.h5ad`
  - Example: `cells.h5ad` with revision 3 → `cells-r3.h5ad`
- **Draft** (`published_at IS NULL`): `{name}-r{revision}-wip-{wip}.h5ad`
  - Example: `cells.h5ad` with revision 1, wip 2 → `cells-r1-wip-2.h5ad`

Where `{name}` is `base_filename` with the extension stripped.

Ingest manifests are excluded — they are already non-downloadable.

---

### Changes

**1. New query: get file download info** (`app/data/files.ts`)

`getFileKey` currently fetches only `key` from `hat.files`. Add a new function `getFileDownloadInfo` that returns everything needed to construct both the S3 request and the download filename:

```sql
SELECT
  f.key,
  f.file_type,
  c.base_filename,
  COALESCE(sd.wip_number, ca.wip_number) AS wip_number,
  COALESCE(sd.published_at, ca.published_at) AS published_at
FROM hat.files f
JOIN hat.concepts c ON c.id = f.concept_id
LEFT JOIN hat.source_datasets sd ON sd.file_id = f.id
LEFT JOIN hat.component_atlases ca ON ca.file_id = f.id
WHERE f.id = $1
```

Also fetch the atlas revision:

```sql
SELECT revision FROM hat.atlases WHERE id = $1
```

Return type:

```typescript
interface FileDownloadInfo {
  atlasRevision: number;
  baseFilename: string;
  key: string;
  publishedAt: Date | null;
  wipNumber: number;
}
```

**2. New helper: build versioned filename** (`app/services/files.ts`)

```typescript
function buildVersionedFilename(
  baseFilename: string,
  revision: number,
  wipNumber: number,
  publishedAt: Date | null,
): string {
  const extIndex = baseFilename.lastIndexOf(".");
  const name = extIndex >= 0 ? baseFilename.slice(0, extIndex) : baseFilename;
  const ext = extIndex >= 0 ? baseFilename.slice(extIndex) : "";
  const versionSuffix = publishedAt
    ? `-r${revision}`
    : `-r${revision}-wip-${wipNumber}`;
  return `${name}${versionSuffix}${ext}`;
}
```

**3. Update `getDownloadUrl`** (`app/services/s3-operations.ts`)

Change signature to accept an explicit download filename:

```typescript
// Before
export async function getDownloadUrl(key: string): Promise<string>

// After
export async function getDownloadUrl(key: string, downloadFilename: string): Promise<string>
```

Use `downloadFilename` in `ResponseContentDisposition` instead of parsing it from the key. The `parseS3KeyPath` call for filename extraction can be removed from this function.

**4. Update `getAtlasFileDownloadUrl`** (`app/services/files.ts`)

Update to use the new query and filename builder, and return both URL and filename:

```typescript
export async function getAtlasFileDownloadUrl(
  atlasId: string,
  fileId: string,
): Promise<PresignedUrlInfo> {
  await confirmFileExistsOnAtlas(fileId, atlasId);
  const info = await getFileDownloadInfo(fileId, atlasId);
  const fileName = buildVersionedFilename(
    info.baseFilename,
    info.atlasRevision,
    info.wipNumber,
    info.publishedAt,
  );
  return {
    fileName,
    url: await getDownloadUrl(info.key, fileName),
  };
}
```

**5. Update `PresignedUrlInfo` type** (`app/apis/catalog/hca-atlas-tracker/common/entities.ts`)

```typescript
export interface PresignedUrlInfo {
  fileName: string;
  url: string;
}
```

### Files to modify

| File | Change |
|------|--------|
| `app/data/files.ts` | Add `getFileDownloadInfo` query |
| `app/services/files.ts` | Add `buildVersionedFilename`, update `getAtlasFileDownloadUrl` |
| `app/services/s3-operations.ts` | Add `downloadFilename` param to `getDownloadUrl` |
| `app/apis/catalog/hca-atlas-tracker/common/entities.ts` | Add `fileName` to `PresignedUrlInfo` |
| `app/services/__mocks__/s3-operations.ts` | Update mock signature |

### Tests

Update `__tests__/api-atlases-id-files-id-presigned-url.test.ts`:

- Verify response includes `fileName` with correct versioned format
- Test published file → `{name}-r{revision}.h5ad`
- Test draft file → `{name}-r{revision}-wip-{wip}.h5ad`
- Unit test `buildVersionedFilename` with edge cases (no extension, multi-dot filenames)

### Dependencies

- `published_at` column must exist on `hat.source_datasets` and `hat.component_atlases` (from a prior slice)

---

### Changes

**1. Use API-returned filename in dialog**

The download dialog (`app/components/.../FileDownloadCell/components/Dialog/dialog.tsx`) currently receives `fileName` as a prop from the table row data. Update it to prefer the `fileName` returned by the presigned URL API response.

- `useRequestPreSignedURL` hook already fetches `PresignedUrlInfo` — extract `fileName` alongside `url`
- Display the API-returned `fileName` in the dialog's "Download Details" section
- Fall back to the prop-provided `fileName` while the API request is in flight

**2. Update hook return type**

`useRequestPreSignedURL` (`app/components/.../FileDownloadCell/hooks/UseRequestPreSignedURL/hook.ts`):

```typescript
// Before
return { url };

// After
return { fileName, url };
```

**3. Update Dialog component**

```typescript
// Use API-returned fileName, fall back to prop
const { fileName: versionedFileName, url } = useRequestPreSignedURL({ fileId });
// ...
<span>FileName: {versionedFileName ?? fileName}</span>
```

### Files to modify

| File | Change |
|------|--------|
| `.../FileDownloadCell/hooks/UseRequestPreSignedURL/hook.ts` | Return `fileName` from API response |
| `.../FileDownloadCell/components/Dialog/dialog.tsx` | Use API-returned filename |
| `.../FileDownloadCell/components/Dialog/entities.ts` | Update props if needed |

---

## Out of scope

- Changing the S3 key structure (files keep their existing keys)
- Versioned filenames for ingest manifests (non-downloadable)
- Filename display in file list tables (separate concern)

### PR DESCRIPTION
**Pull request by hunterckx**
Closes #1115

### CODE REVIEW

File: app/utils/api-handler.ts [hunk 1]
```diff
@@ -271,6 +272,8 @@ function respondError(res: NextApiResponse, error: unknown): void {
     res.status(409).json({ message: error.message });
   else if (error instanceof InvalidOperationError)
     res.status(400).json({ message: error.message });
+  else if (error instanceof S3KeyFormatError)
+    res.status(400).json({ message: error.message });
```
Comment (hunterckx) [inline app/utils/api-handler.ts]: There were some `InvalidOperationError`s being used outside of a strictly API-related context, so I figured I'd give them their own error type (mostly motivated by the `InvalidOperationError` import from `api-handler.ts` causing issues on the frontend; an alternative solution potentially would've been to move the API error types somewhere else)
